pkg/apis/core/validation/validation.go

```diff
@@ -4283,6 +4289,76 @@ func ValidatePodSpec(spec *core.PodSpec, podMeta *metav1.ObjectMeta, fldPath *fi
 	return allErrs
 }
 
+func validatePodResources(spec *core.PodSpec, podClaimNames sets.Set[string], fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
+	if spec.Resources == nil {
+		return nil
+	}
+
+	allErrs := field.ErrorList{}
+	resourceOpts := ResourceValidationOptions{
+		AllowIndivisibleHugePagesValues: opts.AllowIndivisibleHugePagesValues,
+		PodLevelResource:                true,
+	}
+
+	// ValidateResourceRequirements checks if resource names and quantities are
+	// valid, and requests are less than limits.
+	allErrs = append(allErrs, ValidateResourceRequirements(spec.Resources, podClaimNames, fldPath, resourceOpts)...)
+	allErrs = append(allErrs, validatePodResourceConsistency(spec, fldPath)...)
+	return allErrs
+}
+
+// validatePodResourceConsistency checks if effective container-level requests are
+// less than or equal to pod-level requests, and individual container-level limits
+// are less than or equal to pod-level limits.
+func validatePodResourceConsistency(spec *core.PodSpec, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	v1PodSpec := &v1.PodSpec{}
+	if err := corev1.Convert_core_PodSpec_To_v1_PodSpec(spec, v1PodSpec, nil); err != nil {
+		allErrs = append(allErrs, field.InternalError(fldPath, fmt.Errorf("invalid %q: %v", fldPath, err.Error())))
+	}
+
+	reqPath := fldPath.Child("requests")
+	// resourcehelper.PodRequests method requires a Pod object to calculate the total
+	// requests requirements of a pod. Hence a local copy of Pod object is created from PodSpec to
+	// facilitate proper resource calculation without modifying PodRequests method.
+	podReqsFromContainers := resourcehelper.PodRequests(&v1.Pod{Spec: *v1PodSpec}, resourcehelper.PodResourcesOptions{
+		InPlacePodVerticalScalingEnabled: utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling),
```

Comment:
I am not sure if we want to guard counting depending on if feature gates are enabled or not. Otherwise you may end up in a situation where an existing Pod could becomes invalid when you enable a feature gate.

Reply:
This is highlighted in the KEP that a running pod will continue to run, but will become invalid on toggling the feature gate